### ISSUE TITLE
How to speed up the testing process?

### ISSUE THREAD

**Issue by fringe-k** [open-mmlab/mmaction2#2021]
> Before raising a question, you may need to check the following listed items.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://mmaction2.readthedocs.io/en/latest/faq.html) but cannot get the expected help.

Dear all,
       I have my own dataset of about 700 videos and I want to make action detection on them.
       The model I used is **swin_base_patch244_window877_kinetics600_22k.pth**, config is **configs/recognition/swin/swin_base_patch244_window877_kinetics600_22k.py**. 
       I followed demo and my code is here:
   
          device = torch.device(args.device)
            cfg = Config.fromfile(args.config)
            cfg.merge_from_dict(args.cfg_options)
            model = init_recognizer(
                cfg, args.checkpoint, device=device, use_frames=args.use_frames)

          cfg = model.cfg
          device = next(model.parameters()).device  # model device

          # construct label map
          with open(args.label, 'r') as f:
              label = [line.strip() for line in f]

          # build the data pipeline
          test_pipeline = cfg.data.test.pipeline
          test_pipeline = Compose(test_pipeline)
      
          # model = build_model(cfg.model, train_cfg=cfg.get('train_cfg'), test_cfg=cfg.get('test_cfg'))
      
          split="train"
          cfg.data.test.type = 'VideoDataset'
          cfg.data.test.ann_file = f'xxx'
          cfg.data.test.data_prefix = f'xxx'
      
          dataset = build_dataset(cfg.data.test, dict(test_mode=True))
      
          data_loader = build_dataloader(
                  dataset,
                  num_gpus=1,
                  videos_per_gpu=1,
                  workers_per_gpu=cfg.data.workers_per_gpu,
                  dist=False,
                  shuffle=False)
          model = MMDataParallel(model, device_ids=[0])

          labels=[]
          for idx,data in enumerate(tqdm(data_loader)):
              with torch.no_grad():
         
                  scores = model(return_loss=False, **data)[0]
      
                  score_tuples = tuple(zip(label, scores))
                  score_sorted = sorted(score_tuples, key=itemgetter(1), reverse=True)
      
                  top5_label = score_sorted[:5]
                  labels.append(top5_label)

  It works well on 1 V-100 GPU and I got all action labels for this 700 videos. But here's my problem: **the speed of generation is slow like** 
![image](https://user-images.githubusercontent.com/48956356/198309626-65869720-5c72-4015-989f-47f2ca787cee.png)

Now I have a larger dataset containing about 1,000,000 videos and if I want to do same testing things, it will cost me **hundreds of hours**. 
I have tried add "videos_per_gpu"  but got  'max_testing_views is only compatible with batch_size == 1'.

Do you guys have any ideas to **speed up testing**? I will really appreciate it.

**Comment by fringe-k** [comment]
> 
I've tried disabling max_testing_views but found the time is almost the same. Is that normal? Or Dose it mean that the bottleneck is video preprocessing?

**Comment by hukkai** [comment]
Did you try disabling max_testing_views and use a larger batch size?

**Comment by fringe-k** [comment]
> 

Yeah, I disabled max_testing_views and set videos_per_gpu=2 in data_loader. It turns out to be
![image](https://user-images.githubusercontent.com/48956356/198480528-d8133d27-08d6-4026-afb5-a34a5a1e4f32.png)
 
The time is still about 15 mins, and a larger videos_per_gpu > 2 will cause OOM. It's still a little confusing to me. Besides spliting the test set, is there any other ways to further speed up the test? Really thanks for your reply.

**Comment by Dai-Wenxun** [comment]
Maybe you can use val pipeline instead of test pipeline.

**Comment by hukkai** [comment]
If you have that large amount of data, you may use a smaller model say slowonly.

**Comment by hukkai** [comment]
@fringe-k If you have any further questions, feel free to re-open the issue. Thanks!

**Comment by Streamingo-dev** [comment]
> > 
> 
> Yeah, I disabled max_testing_views and set videos_per_gpu=2 in data_loader. It turns out to be ![image](https://user-images.githubusercontent.com/48956356/198480528-d8133d27-08d6-4026-afb5-a34a5a1e4f32.png)
> 
> The time is still about 15 mins, and a larger videos_per_gpu > 2 will cause OOM. It's still a little confusing to me. Besides spliting the test set, is there any other ways to further speed up the test? Really thanks for your reply. All experiments are on a V-100 GPU.

hi @fringe-k 
how did you disable max_testing_views?
when i set it to None, it shows the error(assert isinstance(self.max_testing_views, int) )

Also, I am using a100's and during training i set videos_per_gpu =6,
so did you use the same number of videos_per_gpu in test_cfg as in train_cfg?
